src/rust/engine/src/nodes.rs

```diff
@@ -52,7 +53,10 @@ fn was_required(failure: Failure) -> Failure {
   }
 }
 
-trait GetNode {
+// TODO: This shouldn't be public (is for context to implement GetFileDigest).
+// How should this *actually* be consumed, in such a way that it can be delegated to by traits which
+// have multiple implementations (either for tests, or for weird tools like fs_util).
+pub trait GetNode {
```

Comment:
I'm really not sure why this is a trait anyway... just an artifact maybe? I think you can just move this method to `impl Context`.

Reply:
Done in https://github.com/pantsbuild/pants/pull/5123